onnxruntime/core/providers/nnapi/nnapi_builtin/nnapi_api_helper.cc

```diff
@@ -0,0 +1,120 @@
+
+namespace onnxruntime {
+namespace nnapi {
+
+int32_t GetNNAPIRuntimeFeatureLevel(const ::NnApi* nnapi) {
+  if (!nnapi)
+    return 0;
+  int32_t runtime_level = static_cast<int32_t>(nnapi->nnapi_runtime_feature_level);
+#ifdef __ANDROID__
+  int device_api_level = android_get_device_api_level();
+  runtime_level = (device_api_level < __ANDROID_API_S__) ? device_api_level : runtime_level;
+#endif
+  return runtime_level;
+}
+
+int32_t GetDeviceFeatureLevel(const ::NnApi* nnapi, const std::vector<ANeuralNetworksDevice*>& device_handles) {
+  if (!nnapi)
+    return 0;
+
+  int32_t target_feature_level = GetNNAPIRuntimeFeatureLevel(nnapi);
+
+  int64_t devices_feature_level = -1;
+
+  for (const auto* device_handle : device_handles) {
+    int64_t curr_device_feature_level = 0;
+    if (nnapi->ANeuralNetworksDevice_getFeatureLevel(device_handle, &curr_device_feature_level) != ANEURALNETWORKS_NO_ERROR) {
+      continue;
+    }
+
+    devices_feature_level = std::max(curr_device_feature_level, devices_feature_level);
+  }
+
+  if ((devices_feature_level > 0) && (devices_feature_level < target_feature_level)) {
+    LOGS_DEFAULT(INFO) << "Changing NNAPI Feature Level " << target_feature_level
+                       << " to supported by target devices: " << devices_feature_level;
+
+    target_feature_level = static_cast<int32_t>(devices_feature_level);
+  }
+  return target_feature_level;
+}
+
+[[nodiscard]] Status GetTargetDevices(const ::NnApi* nnapi, TargetDeviceOption target_device_option,
+                                      std::vector<ANeuralNetworksDevice*>& nnapi_target_devices, std::string& nnapi_target_devices_detail) {
+  // GetTargetDevices is only supported on API 29+
+  // get runtime_feature_level
+  if (GetNNAPIRuntimeFeatureLevel(nnapi) < ANEURALNETWORKS_FEATURE_LEVEL_3)
+    return Status::OK();
+
+  const std::string nnapi_cpu("nnapi-reference");
+  uint32_t num_devices = 0;
+  RETURN_STATUS_ON_ERROR_WITH_NOTE(
+      nnapi->ANeuralNetworks_getDeviceCount(&num_devices), "Getting count of available devices");
+
+  int32_t cpu_index = -1;
+  for (uint32_t i = 0; i < num_devices; i++) {
+    ANeuralNetworksDevice* device = nullptr;
+    const char* device_name = nullptr;
+    int32_t device_type = 0;
+    RETURN_STATUS_ON_ERROR_WITH_NOTE(
+        nnapi->ANeuralNetworks_getDevice(i, &device), "Getting " + std::to_string(i) + "th device");
+
+    RETURN_STATUS_ON_ERROR_WITH_NOTE(nnapi->ANeuralNetworksDevice_getName(device, &device_name),
+                                     "Getting " + std::to_string(i) + "th device's name");
+
+    RETURN_STATUS_ON_ERROR_WITH_NOTE(nnapi->ANeuralNetworksDevice_getType(device, &device_type),
+                                     "Getting " + std::to_string(i) + "th device's type");
+    bool device_is_cpu = nnapi_cpu == device_name;
+    if ((target_device_option == TargetDeviceOption::CPU_DISABLED ||
+         target_device_option == TargetDeviceOption::CPU_DISABLED_SOFT) && device_is_cpu) {
```

Comment:
not quite following this - should it be `(CPU_DISABLED || CPU_DISABLED_SOFT) && !device_is_cpu`? also where is `CPU_ONLY` handled now?

Reply:
`CPU_DISABLED_SOFT` should be treated as the same with `CPU_DISABLED` except the 'soft' one allows fallback to CPU if not all ops assigned to the other devices.